compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/QueryTypeChecker.java

```diff
@@ -322,89 +323,162 @@ public BType resolveQueryType(SymbolEnv env, BLangExpression selectExp, BType ta
+                                Types.IntersectionContext.typeTestIntersectionCalculationContext(),
+                                selectType, inferredSelectType, env);
+                    } else {
+                        BType checkedType = symTable.semanticError;
+                        if (selectType == symTable.semanticError) {
+                            checkedType = checkExpr(selectExp, env, data);
+                        }
+
+                        if (checkedType != symTable.semanticError && expTypes.size() == 1) {
+                            selectType = checkedType;
+                        }
+                    }
+
+                    if (queryExpr.isMap) { // A query-expr that constructs a mapping must start with the map keyword.
+                        resolvedType = symTable.mapType;
+                    } else {
+                        resolvedType = getNonContextualQueryType(selectType, collectionType);
+                    }
+                    break;
+            }
+            if (selectType != symTable.semanticError) {
+
+                if (resolvedType.tag == TypeTags.STREAM) {
+                    queryExpr.isStream = true;
                 }
-                if (queryExpr.isMap) { // A query-expr that constructs a mapping must start with the map keyword.
-                    resolvedType = symTable.mapType;
-                } else {
-                    resolvedType = getNonContextualQueryType(selectType, collectionType);
+                if (resolvedType.tag == TypeTags.TABLE) {
+                    queryExpr.isTable = true;
                 }
-                break;
-        }
-        if (selectType != symTable.semanticError) {
-            if (resolvedType.tag == TypeTags.STREAM) {
-                queryExpr.isStream = true;
+                possibleSelectTypes.add(selectType);
+                possibleResolvedTypes.add(resolvedType);
             }
-            if (resolvedType.tag == TypeTags.TABLE) {
-                queryExpr.isTable = true;
+        }
+        if (!possibleSelectTypes.isEmpty() && !possibleResolvedTypes.isEmpty()) {
+            selectTypes.addAll(possibleSelectTypes);
+            resolvedTypes.addAll(possibleResolvedTypes);
+            return;
+        }
+        if (!errorTypes.isEmpty()) {
+            if (errorTypes.size() > 1) {
+                BType actualQueryType = silentTypeCheckExpr(queryExpr, symTable.noType, data);
+                if (actualQueryType != symTable.semanticError) {
+                    types.checkType(queryExpr, actualQueryType, 
+                            BUnionType.create(null, new LinkedHashSet<>(expTypes)));
+                    errorTypes.forEach(expType -> {
+                        if (expType.tag == TypeTags.UNION) {
+                            checkExpr(nodeCloner.cloneNode(selectExp), env, expType, data);
+                        }
+                    });
+                    selectExp.typeChecked = true;
```

Comment:
At this level selectExp hasn't been typeChecked isn't it? Is it correct to mark it as type-checked?

Reply:
Yes. There are case that can be where selectExpr not being type checked.